timemachine/fe/mcgregor.py

```diff
@@ -135,45 +126,105 @@ def __init__(self, n_vertices, edges):
         2. M-D-M # returns True
         3. M-M-U # returns False
         4. M-M-D # returns False
 
-        The implementation is adapted from the _plain_bfs() method in:
-        https://networkx.org/documentation/stable/_modules/networkx/algorithms/components/connected.html
-
+        This calls an optimized implementation in mapping_is_disconnected_fast; see mapping_is_disconnected_ref for a
+        simpler-to-understand reference version.
         """
-        unseen = [True] * self.n_vertices
-        unseen[source] = False
-        nextlevel = [source]
-        mapped_count = 1
-
-        if mapped_count == n_mapped_nodes:
-            return False
-
-        while nextlevel:
-            thislevel = nextlevel
-            nextlevel = []
-            for v in thislevel:
-                for w in self.get_neighbors(v):
-                    if unseen[w] and node_states[w] != NODE_STATE_VISITED_DEMAPPED:
-                        if node_states[w] == NODE_STATE_VISITED_MAPPED:
-                            mapped_count += 1
-                            if mapped_count == n_mapped_nodes:
-                                return False
-                        unseen[w] = False
-                        nextlevel.append(w)
-        return True
+
+        # Uncomment to assert result of fast implementation matches the reference
+        # args = (mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size)
+        # assert self.mapping_is_disconnected_fast(*args) == self.mapping_is_disconnected_ref(*args)
+
+        return self.mapping_is_disconnected_fast(
+            mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size
+        )
+
+    def mapping_is_disconnected_fast(
+        self,
+        mapped_nodes: Set[int],
+        unvisited_nodes: Set[int],
+        max_connected_components: Optional[int],
+        min_connected_component_size: int,
+    ):
+        """Optimized implementation of mapping_is_disconnected, sacrificing modularity for efficiency"""
+
+        seen = set()
+        n_ccs = 0
+        for u in mapped_nodes:
+            if u not in seen:
+                # visit component containing v
+                seen.add(u)
+                cc_size = 1
+                nextlevel = [u]
+                while nextlevel:
+                    thislevel = nextlevel
+                    nextlevel = []
+                    for v in thislevel:
+                        for w in self.get_neighbors(v):
+                            if w in mapped_nodes or w in unvisited_nodes:
+                                if w not in seen:
+                                    seen.add(w)
+                                    cc_size += 1
+                                    nextlevel.append(w)
+                n_ccs += 1
+                if cc_size < min_connected_component_size:
+                    return True
+                if n_ccs == max_connected_components:
```

Comment:
In `_ref` version there's a short circuit for `None` - does this still hold up for the `_fast` version?

Reply:
To clarify, in the event that `n_ccs=0` and `max_connected_components=None` this would lead to `0 == None`, which is `False` - I *think* this is as intended (i.e. bypassing the check completely).

Reply:
The short-circuit is [here](https://github.com/proteneer/timemachine/blob/3cb2a8d63f8b4b817119a5872790a9c819ee8f14/timemachine/fe/mcgregor.py#L435) and applies to either implementation in the case where `max_connected_components is None and min_connected_component_size <= 1`. (Even if `max_connected_components is None`, we may still have to look at the connected components to respect the minimum size filter.)

Reply:
> To clarify, in the event that n_ccs=0 and max_connected_components=None this would lead to 0 == None, which is False - I think this is as intended (i.e. bypassing the check completely).

Yes, that is intentional, though admittedly confusing. I went back and forth on whether to explicitly handle `None` there; can put it back because I doubt there's any performance impact to being explicit.

Reply:
Opted to handle `None` explicitly in https://github.com/proteneer/timemachine/pull/1349/commits/38b29a4a8ee3011ee160c991e235a292445980ba